### PR TITLE
Add `transform_colorspace_to_srgb` operation and use it to fix inaccurate colors when saving specific image files

### PR DESCRIPTION
**Pull request by Stormheg**
Please see https://github.com/wagtail/Willow/issues/139 for a detailed description of what this is trying to accomplish. Took me longer than expected to prepare a PR, writing tests turned out to be quite time consuming!

I've expanded the automatic conversion discussed in #139 to include `png` files too as they are also affected by the color issue described in #139. Technically, GIF files are also affected but I've not bothered to fix them because there isn't much to gain when the colors are butchered to fit in a palette of 256 colors.

The automatic conversion during save of `avif`, `heic`, `webp` and `png` images _only_ happens when we encounter an image that is affected by the issue (CMYK mode, with embedded ICC profile). For all other images, the new code path is not taken and there should be no impact.

I'd appreciate reviews from anyone! (not necessarily Willow maintainers only).

I've done my best to explain how and why in the docs and comments but I'm no expert on the subject of color spaces and gamuts and such. Please let me know if there are any inaccuracies or ambiguities.

### CODE REVIEW

File: tests/test_pillow.py [hunk 1]
```diff
@@ -124,6 +125,72 @@ def test_set_background_color_rgb_color_argument_check(self):
...
+        # deviations from the expected result without being too sensitive to minor
+        # differences between systems.
+        stat = ImageStat.Stat(image_srgb.image)
+        expected_sum2 = [1286116001.0, 1222827100.0, 1144012271.0]
+
+        for actual, expected in zip(stat.sum2, expected_sum2):
+            self.assertAlmostEqual(
+                actual,
+                expected,
+                delta=1000,
```
Comment (Stormheg) [inline tests/test_pillow.py:185]: I've assumed that the results on my system might not be the same as on other systems. Hence the the `assertAlmostEqual`.

I haven't actually run this test on any other systems so not sure if my assumption is actually true.

**Comment by Stormheg** [comment]
I've added a new image for the test suite: `dog_and_lake_cmyk_with_icc_profile.jpg`. I've modified this image specifically to be affected. The image has been heavily downsized to make its size as small as possible but the embedded ICC profile still takes up a lot of space. It is a little over 1MB in size.

I took the image a couple years ago during a hike with my dog. Willow has my permission to use the downsized image file for its test suite.

Here is a full-sized image for your viewing pleasure. Try opening and saving it as `webp` or `avif` with an unpatched version of Willow and this patch to see the difference.
[zaar-lake_iso_coatedv2.jpg.zip](https://github.com/wagtail/Willow/files/13932094/zaar-lake_iso_coatedv2.jpg.zip)

image

_screenshot comparing the original JPEG on the left with an unpatched WEBP version._

File: willow/plugins/pillow.py [hunk 2]
```diff
@@ -227,19 +278,32 @@ def save_as_png(self, f, optimize: bool = False, apply_optimizers: bool = True):
...
+                pillow_image = self.transform_colorspace_to_srgb()
+                image = pillow_image.image
+                kwargs["icc_profile"] = pillow_image.get_icc_profile()
+            else:
+                kwargs["icc_profile"] = icc_profile
+
         if self.image.mode == "CMYK":
             image = self.image.convert("RGB")
         else:
             image = self.image
```
Comment (zerolab) [inline willow/plugins/pillow.py:302]: This now does double duty. I believe this now need to be

```suggestion
        elif self.image.mode == "CMYK":
            # we don't have an ICC profile, and the image is in CMYK, so convert to RGB
            image = self.image.convert("RGB")
```

File: tests/test_pillow.py (hunk 1 above)
Comment (zerolab) [reply tests/test_pillow.py:185]: Let's try to be more strict and see what the CI says

**Review by zerolab — COMMENTED** [review]
Nice one @Stormheg 
Left a couple of notes

**Comment by andre-fuchs** [comment]
@Stormheg Very cool! I suggest to add a system wide setting to convert all images to sRGB. Two setting variables would desirable, I think. A boolean type to turn this conversion on or off. And another one for the rendering intend (could be both an integer or char type). This is mainly an update of Wagtail, but I guess that Willow also would need some preparation for that. Does this make sense?

**Comment by Stormheg** [comment]
@andre-fuchs lets discuss automatically converting to sRGB in a separate issue 👍 

Since most images and devices consuming the web are only capable of displaying sRGB this should be mostly fine; the space saving would be worth the cost in most cases I think. I could imagine that a photography site might not want to perform such conversions however. Saturated sRGB colors look a bit bland compared to what is possible with P3 and Adobe RGB.

Maybe there could be a middle ground where we provide functionality to detect what color gamut the original image is in? And if it is in a color other than sRGB, Wagtail could output something akin to the snippet below. This will only load the wide gamut with original ICC profile for devices that support displaying that color gamut.

```html
<picture>
  <source media="(color-gamut: p3)" srcset="photo-wide.jpg">
  <img src="photo-srgb.jpg">
</picture>
```

Somewhat niche functionality maybe, but with the number of devices that are capable of recording and displaying wide-gamut images increasing (in particular recent iPhones are capable of this) it would be good future proofing for Wagtail to be aware of wide gamut images.

Anyway, let's discuss separately from this PR 😄

File: willow/plugins/pillow.py (hunk 2 above)
Comment (Stormheg) [reply willow/plugins/pillow.py]: Little oversight of mine, thanks!

File: tests/test_pillow.py (hunk 1 above)
Comment (Stormheg) [reply tests/test_pillow.py]: CI is green with no delta, I'll make this much simpler.

**Review by zerolab — APPROVED** [review]

**Comment by zerolab** [comment]
Nice one @Stormheg. Let's discuss the suggestion from @andre-fuchs separately

**Pull request merged**